docs/design-documents/drivers/serial/serial.md

```diff
@@ -0,0 +1,321 @@
+
+**API description**
+
+```C
+class MinimalConsole : public MinimalSerial, private NonCopyable<MinimalConsole> {
+public:
+    MinimalConsole(PinName tx, PinName rx, int baud);
+    virtual ~MinimalConsole();
+    int getc();
+    int putc(int c);
+};
+```
+
+The constructor provides arguments to create a serial port connected to the specified transmit and receive pins and set to the specified baud rate. If a target has serial support, by default a serial port is used for the console and the pins and settings for the port selection come from target header files and JSON settings. 
+
+# Usage scenarios and examples
+
+### Scenario 1 `Printing to stdout`
+
+The below example shows pseudocode for how to print `Hello` to stdout using the new API. Printf cannot be called directly; instead the application first calls _fdopen_ to get a _FILE*_ to use the printf function.
+
+```C
+BufferedSerial serial(RX, TX, 115200);
+FILE *out = fdopen(&serial, "w");
+fprintf(out, "Hello");
+```
+and similarly for `UnbufferedSerial`
+
+```C
+UnbufferedSerial serial(RX, TX, 115200);
+FILE *out = fdopen(&serial, "w");
+fprintf(out, "Hello");
+```
+
+# Other Work
+
+## Related work-Simplification of retarget code
+
+In Mbed, retargeting is implemented by redefining the system I/O functions.
+The `FileHandle` abstract class provides an interface with file-like operations, such as read and write that can be redefined by the retarget code. The high-level library functions perform input/output by calling the low-level functions which use the system I/O functions to interface with hardware.
+
+The relationship of `FileHandle` and other APIs is as follows:
+
+![Retarget](./diagrams/retarget.png)
+
+The POSIX layer and the static array of `FileHandle*` allocated by the retarget code for file handling consume a lot of memory. A constrained target that only uses a console does not need file handling and should be able to turn this functionality off.
+
+ A new configuration parameter is introduced to compile the file handling code out. The application can override the default configuration in mbed_app.json as follows:
+
+```C
+    "target_overrides": {
+        "*": {
+            "platform.stdio-minimal-console-only": true
+        }
+    }
+```
+
+There are two approaches to achieve a minimal retarget layer. You can:
+1. Redefine the low-level library functions. A detailed description can be found [here](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0475m/chr1358938931411.html)
+1. Redefine the system I/O functions.
+
+A minimal console only needs to write a single character at a time. So redefining the default fputc() to directly write to the serial port if the output stream is stdout and bypassing the system I/O functions should achieve higher memory savings. If we take this approach, we will have to rework some error handlers that rely on the POSIX form of `write(STDOUT_FILENO, buf, len)` to do emergency printing.
```

Comment:
That method is also toolchain dependent - I don't know if we've established how easy it is with the other 2.

Reply:
Indeed the first approach looks more complex. We should only pursue if it brings more memory saving. We'll decide during implementation.